### PR TITLE
Generate and send weekly outcomes reports to the clients every Monday morning

### PR DESCRIPTION
**Pull request by g-vega-cl**
## Purpose  
- To generate and send weekly outcomes reports to the clients every Monday morning.
- Every Monday 11 A.M messages.api.ts runs `weeklyReport()` . WeeklyReport then finds all patients that have outcomes from last week. Then, for each patient it calculates the data required by the message template and creates the templates. Then it creates the message and saves it so the Twilio scheduler can send the message.

## Changes and Additions 
Added `supertest` for integration tests.
`BackgroundJobs.integration.test.ts` tests:
 - `runCronSchedules() `schedules get called at the appropiate times.
 - sends midnight messages with `dailyMidnightMessages()`
 - The first time `weeklyReport()`runs, it waits until the next monday to run.
 - `weeklyReport()` Only runs once per week
 - `weeklyReport()` sends last weeks messages from monday to sunday, regardles of the day of the week it runs.
 - Test only the most recent message of a day is parsed

created `schedule.model.ts` so we can store the last time the weekly report was sent.

`scheduling.integration.tests.ts` tests that scheduled messages are sent at appropriate times.

`scheduling` now sends message backlog even if the server gets stuck. (It sends them when the server goes back online).

`db.test.ts`tests that we can use supertest to get an authorization token for tests.

`db.ts` created functions to mock databases with to use with jest tests in development.

`outcome.model.ts` Changed patientID type to string, and made alert types have to match the types in the glucoseParser. 
		 
 In `messageTemplate.api.ts` deleted `if (req.body.image == null)` and its contents because the exact same operation was repeated below. 

 moved logic from `messages.api.ts` to `utils.ts`

 `utils.ts` creates and sends the scheduled daily messages  `dailyMidnightMessages()` and the weekly sugar report `weeklyReport()`

`messages.api.ts` runs the `cron` jobs for the daily and weekly messages.

## Open Trello Tickets

-[[Reports] Generate and send weekly outcomes report to each client every Monday morning](https://trello.com/c/XSKIQAR3/72-reports-generate-and-send-weekly-outcomes-report-to-each-client-every-monday-morning)

## Tested
- Branch is up to date to Master Branch
- Ran `yarn lint` before MR
- Ran `yarn test` before MR
- Application runs without crashes (`yarn start`)

Test coverage:
Screen Shot 2021-06-15 at 17 14 37

Logic: 
Week 0 - The clients send messages to the glucose number with their readings. These readings get parsed and stored in an outcomes database.
 - Edge case: If the user sends more than one message in a day, we take the value of the most recent message.
 - Edge case: If we can't parse the user message, the outcome is not saved and the user receives an error message.
Week 1 to ∞ - Monday 11 A.M PST, the database creates the weekly report with the outcomes data and sends it to their respective patients.
 - Edge case: If the server is not working on Monday 11 A.M, the messages will be sent on Tuesday, if it's not available on Tuesday, they'll be sent on Wednesday, etc...
 - Edge case: When the code is pushed for the first time, the automated outcomes will be sent until the next Monday ay (If you push it on Tuesday, the first reports will be sent until Monday 11 A.M).

### CODE REVIEW

File: src/models/outcome.model.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ const { Schema } = mongoose;
 
 interface IOutcome extends mongoose.Document {
   _id: string;
-  patientID: number;
+  patientID: string;
```
Comment (rweald) [inline src/models/outcome.model.ts:7]: Why the need for a type change here? Seems like it might have some unintended consequences upstream.

File: src/routes/messages/cronSchedules.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { dailyMidnightMessages, weeklyReport } from './utils';
```
Comment (rweald) [inline src/routes/messages/cronSchedules.ts:1]: If we are already going to extract this to a new file I think we should go one step further and move it out of the routes namespace. It doesn't actually have anything to do with routing. 

As we add more and more background jobs it makes much more sense to have them all colocated in something like `src/backgound_jobs`

File: src/routes/messages/utils.ts [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+  weekAverage: number,
+  weekRecords: IweekRecords,
+  language: string,
+) => {
+  let body = '';
+  if (recordedCount > 0) {
+    const averageColor = returnColorRanges(weekAverage);
+    if (language === 'english') {
+      body += `${averageColor} Average: ${weekAverage} mg/dL\n\n`;
+      if (weekRecords.monday > 0) {
```
Comment (rweald) [inline src/routes/messages/utils.ts:88]: Would be nice to extra this logic into a general function that can take a day as argument. That way we don't need so many if statements that is really just iterating through a fixed list of days.

File: src/routes/messages/utils.ts [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        }\n`;
+      }
+      if (weekRecords.sunday > 0) {
+        body += `  ${returnColorRanges(weekRecords.sunday)} Sun: ${
+          weekRecords.sunday
+        }`;
+      }
+    }
+
+    if (language === 'spanish') {
```
Comment (rweald) [inline src/routes/messages/utils.ts:125]: The above comment about a general function is even more relevant since we duplicate all the if/then statements here as well.

File: src/routes/messages/utils.ts [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+      }
+    }
+  });
+  weekAverage /= recordedCount;
+  weekAverage = Math.round(weekAverage);
+
+  return [weekAverage, recordedCount, greenCount];
+};
+
+export const weeklyReport = () => {
```
Comment (rweald) [inline src/routes/messages/utils.ts:220]: This function is way way way too big. We need to break down the constituent parts into pieces that can be more easily reasoned about and tested.

**Review by rweald — COMMENTED** [review]
1 meta issue I'm noticing is that this job can only run successfully exactly on the first day of the week. If there is any issues with the ops and the job needs to be run later, the logic will fail. 

This might be okay, but it would have been much better to write the logic in such a way it can be run on any day of the week for the entire proceeding week.

File: src/routes/messages/utils.ts [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+            date: {
+              $gt: new Date(
+                today.getFullYear(),
+                today.getMonth(),
+                today.getDate() - 7,
+              ),
+            },
+            patientID: patient._id,
+          })
+            .then((outcomes) => {
```
Comment (rweald) [inline src/routes/messages/utils.ts:237]: Please move this away from `then` to `async/await` as requested in the README

File: src/routes/messages/cronSchedules.ts (hunk 2 above)
Comment (g-vega-cl) [reply src/routes/messages/cronSchedules.ts:1]: Perfect, I moved the cronSchedules.ts code to a new folder in src/background_jobs

File: src/routes/messages/utils.ts (hunk 3 above)
Comment (g-vega-cl) [reply src/routes/messages/utils.ts:88]: Great catch!
I moved the logic to a function and now we don't have the nested ifs!

File: src/models/outcome.model.ts (hunk 1 above)
Comment (g-vega-cl) [reply src/models/outcome.model.ts:7]: This one is interesting. I changed it because the ._id in `patient.model.ts` is a string. and when filtering Outcomes, we get the patientID from the patient model.

Also, when filtering, using new ObjectId(patient._id) to coerce to number does not work, unlike when creating new Outcomes. 

`Outcome.find({patientID: new ObjectId(patient._id)});` <- won't work.
but:
`new Outcome(... ,patientID: new ObjectId(patient._id)...) `<- does work.

What do you think?

File: .github/workflows/staging-ci.yml [hunk 7]
```diff
@@ -0,0 +1,48 @@
+# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
```
Comment (rweald) [inline .github/workflows/staging-ci.yml:1]: We don't really want and/or need this as part of this PR. Let's keep it focused on the pure report generation stuff.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -15,7 +15,7 @@ jobs:
     runs-on: ubuntu-latest
 
     env:
-      TWILIO_ACCOUNT_SID: some_fake_string
+      TWILIO_ACCOUNT_SID: AC-some_fake_string
```
Comment (rweald) [inline .github/workflows/ci.yml:18]: Same as other comment. Is there a reason we need to touch this file?

File: src/background_jobs/utils.ts [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+interface IweekRecords {
+  [char: string]: number;
+}
+
+export const dailyMidnightMessages = () => {
+  console.log('Running batch of scheduled messages');
+  Patient.find().then((patients) => {
+    MessageTemplate.find({ type: 'Initial' })
+      .then((MessageTemplates) => {
```
Comment (rweald) [inline src/background_jobs/utils.ts:18]: Can we move this over to async / await. I noticed we are mixing and matching async/await and then style all in 1 function.

File: src/background_jobs/backgroundJobs.integration.test.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    it('weekly reports run if it has not run this week', async (done) => {
+      const lastSunday = getDateRelativeToMonday(-1);
+      await new Schedule({ weeklyReport: lastSunday }).save();
+      await createPatient('111');
+      const patient = await Patient.findOne({});
+      if (patient) {
+        await createOutcome(patient, getDateRelativeToMonday(1), 97, 'green');
+        await createOutcome(patient, getDateRelativeToMonday(0), 77, '<80');
+        await createOutcome(patient, getDateRelativeToMonday(-1), 99, 'green');
+        await createOutcome(patient, getDateRelativeToMonday(-.6), 111, 'green');
```
Comment (rweald) [inline src/background_jobs/backgroundJobs.integration.test.ts:97]: Mind quickly fixing these couple of linter errors. Should be super easy to run prettier on the file to address.

File: src/background_jobs/backgroundJobs.integration.test.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
+import {
```
Comment (rweald) [inline src/background_jobs/backgroundJobs.integration.test.ts:1]: Can we put this file in the test directory along with all the other tests. There was already an existing convention.

File: src/background_jobs/backgroundJobs.test.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
+import {
```
Comment (rweald) [inline src/background_jobs/backgroundJobs.test.ts:1]: Same as above, please put tests in the `/test` directory. Ensures that the compiled end result that we use in production does not include any tests and was the existing convention.

File: src/background_jobs/utils.ts [hunk 13]
```diff
@@ -0,0 +1,296 @@
+/* eslint-disable no-console */
+/* eslint-disable prettier/prettier */
```
Comment (rweald) [inline src/background_jobs/utils.ts:2]: Don't disable prettier. It should run on all source code files.

File: src/index.ts [hunk 14]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-console */
```
Comment (rweald) [inline src/index.ts:1]: I opened a new PR that just disables the eslint console warnings globally https://github.com/coachmehealth/sms-care-portal-backend/pull/46

As a general practice you should never be disabling eslint rules on a per file basis unless there is a very explicit reason, like working around a eslint or typescript bug.

File: test/db.ts [hunk 15]
```diff
@@ -0,0 +1,146 @@
+/* eslint-disable no-restricted-syntax */
+/* eslint-disable guard-for-in */
```
Comment (rweald) [inline test/db.ts:2]: Again, probably shouldn't be disabling this. ESLint is there for a reason, the code can be changed to avoid the issue the linter is complaining about.

**Review by rweald — APPROVED** [review]
Ready to squash merge.
